File: Homework/Hoang_Nguyen/lesson3/238_productOfArrayExceptSelf.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  const result = [];
  // calculate reverse prefix products
  result[nums.length - 1] = nums[nums.length - 1];
  for (let i = nums.length - 2; i >= 0; i--) {
    result[i] = nums[i] * result[i + 1];
  }
  let prefixProduct = 1;
  let i = 0;
  for (i = 0; i < nums.length - 1; i++) {
    result[i] = prefixProduct * result[i + 1];
    prefixProduct *= nums[i];
  }
  result[i] = prefixProduct;
  return result;
};

File: Homework/quoc_khanh/lesson_2/238.ProductOfArrayExceptSelf.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length = len(nums)
        answer = [1] * (length + 1)
        # calculate suffix products
        prev = 1
        for i in reversed(range(length)):
            answer[i] = prev * nums[i]
            prev = answer[i]
        # calculate result
        cumulativeProduct = 1
        for i in range(length):
            answer[i] = cumulativeProduct * answer[i + 1]
            cumulativeProduct *= nums[i]
        return answer[:-1]

s = Solution()
print(s.productExceptSelf([1,2,3,4]) == [24,12,8,6])
print(s.productExceptSelf([-1,1,0,-3,3]) == [0,0,9,0,0])